stream/stream_bstr.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "stream_bstr.h"
+#include "common/common.h"
+
+struct priv {
+    bstr data;
+};
+
+static int fill_buffer(stream_t *s, void *buffer, int len)
+{
+    struct priv *p = s->priv;
+    bstr data = p->data;
+    if (s->pos < 0 || s->pos > data.len)
+        return 0;
+    len = MPMIN(len, data.len - s->pos);
+    memcpy(buffer, data.start + s->pos, len);
+    return len;
+}
+
+static int seek(stream_t *s, int64_t newpos)
+{
+    return 1;
+}
+
+static int64_t get_size(stream_t *s)
+{
+    struct priv *p = s->priv;
+    return p->data.len;
+}
+
+int stream_bstr_open2(stream_t *stream, const struct stream_open_args *args, bstr data)
+{
+    stream->fill_buffer = fill_buffer;
+    stream->seek = seek;
+    stream->seekable = true;
+    stream->get_size = get_size;
+
+    struct priv *p = talloc_zero(stream, struct priv);
+    stream->priv = p;
+
+    p->data = bstrdup(stream, data);
```

Comment:
if you dup it here again, who frees the original string?

Reply:
Should be the callers responsibility.
We don't know where the original data comes from and if it even can be freed.